test/suite/test_prepare07.py

```diff
@@ -0,0 +1,165 @@
+        session_p.prepare_transaction('prepare_timestamp=' + timestamp_str(130))
+
+        # Commit some more updates.
+        self.session.begin_transaction('isolation=snapshot')
+        cursor.set_key(ds.key(nrows + 4))
+        cursor.set_value(value_b)
+        self.assertEquals(cursor.update(), 0)
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(140))
+        self.session.begin_transaction('isolation=snapshot')
+        cursor.set_key(ds.key(nrows + 5))
+        cursor.set_value(value_b)
+        self.assertEquals(cursor.update(), 0)
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(150))
+
+        # Move the oldest and the stable timestamp to the latest.
+        self.conn.set_timestamp('stable_timestamp=' + timestamp_str(155))
+        self.conn.set_timestamp('oldest_timestamp=' + timestamp_str(155))
+
+        # Commit an update newer to the stable timestamp.
+        self.session.begin_transaction('isolation=snapshot')
+        cursor.set_key(ds.key(nrows + 6))
+        cursor.set_value(value_b)
+        self.assertEquals(cursor.update(), 0)
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(160))
+
+        # Take a checkpoint.
+        self.session.checkpoint()
+
+        # Commit the prepared transaction.
+        session_p.commit_transaction('commit_timestamp=' + timestamp_str(140) + ',durable_timestamp=' + timestamp_str(160))
+
+        # Take a backup and check the values.
+        self.backup_dir = os.path.join(self.home, "WT_BACKUP")
+        self.backup(self.backup_dir)
+        backup_conn = self.wiredtiger_open(self.backup_dir)
+        session_b = backup_conn.open_session()
+        cursor_b = session_b.open_cursor(uri)
+
+        # Committed non - prepared transactions data should be seen.
+        cursor_b.set_key(ds.key(nrows + 1))
+        self.assertEquals(cursor_b.search(), 0)
+        self.assertEquals(cursor_b.get_value(),value_b)
+        cursor_b.set_key(ds.key(nrows + 2))
+        self.assertEquals(cursor_b.search(), 0)
+        self.assertEquals(cursor_b.get_value(),value_b)
+        # Committed prepared transaction data should not be seen.
+        cursor_b.set_key(ds.key(nrows + 3))
+        self.assertEquals(cursor_b.search(), 0)
+        self.assertEquals(cursor_b.get_value(),value_a)
+        # Committed non - prepared transactions data should be seen.
+        cursor_b.set_key(ds.key(nrows + 4))
+        self.assertEquals(cursor_b.search(), 0)
+        self.assertEquals(cursor_b.get_value(),value_b)
+        cursor_b.set_key(ds.key(nrows + 5))
+        self.assertEquals(cursor_b.search(), 0)
+        self.assertEquals(cursor_b.get_value(),value_b)
+
+        # Committed transactions newer to the stable timestamp should not be seen.
+        cursor_b.set_key(ds.key(nrows + 6))
+        self.assertEquals(cursor_b.search(), 0)
+        self.assertEquals(cursor_b.get_value(),value_a)
+
```

Comment:
Adding an additional test scenario of validating the prepared updates after moving the stable timestamp? Is it worth?

Reply:
I think yes, the reason being ```__wt_txn_visible_all``` uses the oldest txn id, and as part of prepare_transaction we release transaction id. The visibility of a prepared transaction is controlled by the timestamp and state rather than txn id, till its resolution. So, when the prepared transaction is the oldest transaction, eviction/checkpoint should not make prepared updates durable based on the txn id, but should consider the timestamp and state.